tpu-client/src/tpu_client.rs

```diff
@@ -95,6 +105,66 @@ where
         self.invoke(self.tpu_client.try_send_transaction(transaction))
     }
 
+    /// Serialize and send transaction to the current and upcoming leader TPUs according to fanout
+    /// size
+    /// Attempts to send and confirm tx "tries" times
+    /// Waits for signature confirmation before returning
+    /// Returns the transaction signature
+    pub fn send_and_confirm_transaction_with_retries<T: Signers + ?Sized>(
+        &self,
+        keypairs: &T,
+        transaction: &mut Transaction,
+        tries: usize,
+        pending_confirmations: usize,
+    ) -> TransportResult<Signature> {
+        for x in 0..tries {
+            let now = Instant::now();
+            let mut num_confirmed = 0;
+            let mut wait_time = MAX_PROCESSING_AGE;
+            // resend the same transaction until the transaction has no chance of succeeding
+            let wire_transaction =
+                bincode::serialize(&transaction).expect("transaction serialization failed");
+
+            while now.elapsed().as_secs() < wait_time as u64 {
+                let leaders = self
+                    .tpu_client
+                    .get_leader_tpu_service()
+                    .leader_tpu_sockets(self.tpu_client.get_fanout_slots());
+                if num_confirmed == 0 {
+                    for tpu_address in &leaders {
+                        let cache = self.tpu_client.get_connection_cache();
+                        let conn = cache.get_connection(tpu_address);
+                        conn.send_data_async(wire_transaction.clone())?;
```

Comment:
It looks like `clone` is not needed

Reply:
i think the `clone` is needed here because `send_data_async()` takes ownership over `wire_transaction`. And we are looping over the leaders and sending `wire_transaction` to the upcoming leaders